### PR TITLE
Add a configurable version option

### PR DESCRIPTION
**Pull request by jaxxstorm**
##### SUMMARY
Instead of hardcoding the URL, you can now simply update the version
string. This also means specifying your own URL is relatively
simple/straightforward

##### TESTS/SPECS

Tests are passing, have added facter facts to some of the tests

### CODE REVIEW

File: manifests/params.pp [hunk 1]
```diff
@@ -58,4 +60,13 @@
       fail("Module ${module_name} is not supported on osfamily '${::osfamily}'")
     }
   }
+	case $::architecture {
```
Comment (jsok) [inline manifests/params.pp]: Indentation seems off here

File: manifests/params.pp [hunk 2]
```diff
@@ -58,4 +60,13 @@
...
+    'x86_64', 'amd64': { $arch = 'amd64' }
+    'i386':            { $arch = '386'   }
+    /^arm.*/:          { $arch = 'arm'   }
+    default:           {
+      fail("Unsupported kernel architecture: ${::architecture}")
+    }
+  }
+	$os = downcase($::kernel)
```
Comment (jsok) [inline manifests/params.pp]: de-indent

File: manifests/init.pp [hunk 3]
```diff
@@ -29,7 +29,10 @@
 #   generated config.
 #
 # * `download_url`
-#   URL to download the vault zip distribution from.
+#   Manual URL to download the vault zip distribution from.
+#
+# * `download_url_base`
```
Comment (jsok) [inline manifests/init.pp:34]: Can you please also update the `README.md` to reflect the new params?

I've looked into using `puppet-strings` for this but it doesn't quite support generating markdown out of the box so I'm just manually keeping things in sync for now.

File: manifests/params.pp [hunk 4]
```diff
@@ -58,4 +60,13 @@
       fail("Module ${module_name} is not supported on osfamily '${::osfamily}'")
     }
   }
+	case $::architecture {
+    'x86_64', 'amd64': { $arch = 'amd64' }
+    'i386':            { $arch = '386'   }
+    /^arm.*/:          { $arch = 'arm'   }
+    default:           {
+      fail("Unsupported kernel architecture: ${::architecture}")
```
Comment (jsok) [inline manifests/params.pp:69]: In the scenario that I made a custom build of vault for an architecture that doesn't exist here and specified it via `download_url` the module wouldn't allow me to install it.

i.e. this test case would fail:

```
let(:facts) {{
  :architecture => 'ppc'
}}
let(:params) {{
  :download_url => 'https://example.com/vault-ppc.zip'
}}
```

Admittedly it's a bit of an edge case but is introduced by this change.

**Review by jsok — CHANGES_REQUESTED** [review]
Thanks, I like this approach since it still allows easily customising the download URL.

I would also appreciate some tests that exercise and assert using a customised base url and version/os/arch combo. At this stage there is no spec to catch a bad `real_download_url` being constructed.

File: manifests/params.pp (hunk 4 above)
Comment (jaxxstorm) [reply manifests/params.pp:69]: Seems reasonable. I'll set $arch to undef by default.

Comment (jaxxstorm) [reply manifests/params.pp:69]: @jsok under further review, the module currently only has support for Debian and Redhat. I know Debian is fairly renowned for supporting lots of different arch/kernels, but sane defaults makes more sense? I am happy to check for the presence of the `download_url` inside the case statement, but I expect the code to get very, very unreadable after a while.

File: manifests/init.pp [hunk 5]
```diff
@@ -108,6 +118,10 @@
     validate_string($max_lease_ttl)
   }
 
+  # lint:ignore:140chars
+  $real_download_url    = pick($download_url, "${download_url_base}${version}/${package_name}_${version}_${os}_${arch}.${download_extension}")
```
Comment (jsok) [inline manifests/init.pp:126]: I just noticed `download_extension` doesn't appear to be defined anywhere...

File: manifests/params.pp (hunk 4 above)
Comment (jsok) [reply manifests/params.pp:69]: That's fine, it's definitely an unlikely scenario so not worth introducing the complexity for it.

File: manifests/init.pp (hunk 5 above)
Comment (jaxxstorm) [reply manifests/init.pp:126]: Ah good catch, will fix.

**Comment by jaxxstorm** [comment]
@jsok I've updated the readme, and added some additional specs and the missing `download_extension` param. Let me know if you need anything else!

File: spec/classes/vault_spec.rb [hunk 6]
```diff
@@ -84,6 +86,29 @@
...
 
+        context "default download options" do
+          it {
+            is_expected.to contain_archive('/tmp/vault.zip')
+              .with_source('https://releases.hashicorp.com/vault/0.6.5/vault_0.6.5_linux_amd64.zip')
+              .that_comes_before('File[/usr/local/bin/vault]')
+          }
+        end
+
+        context "specifying a custom download params" do
```
Comment (jsok) [inline spec/classes/vault_spec.rb:97]: 👍

**Review by jsok — APPROVED** [review]
